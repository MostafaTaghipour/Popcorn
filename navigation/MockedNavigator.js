import React from "react";
import { NavigationContainer } from "@react-navigation/native";
import { createStackNavigator } from "@react-navigation/stack";
import { Provider as ReduxProvider } from "react-redux";
import { AuthProvider } from "../contexts/AuthContext";

const Stack = createStackNavigator();

/**
 * Provide a MockedNavigator to wrap our screens
 * 
 * Access navigation in screens 
 * 
 * @param {*} param0 
 */
export const MockedNavigator = ({ component, params = {}, store }) => {
  return (
    <ReduxProvider store={store}>
      <AuthProvider>
        <NavigationContainer>
          <Stack.Navigator>
            <Stack.Screen
              name="MockedScreen"
              component={component}
              initialParams={params}
            />
          </Stack.Navigator>
        </NavigationContainer>
      </AuthProvider>
    </ReduxProvider>
  );
};
