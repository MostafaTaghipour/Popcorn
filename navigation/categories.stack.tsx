import React from "react";
import {
  createStackNavigator,
  StackNavigationProp,
} from "@react-navigation/stack";
import { RouteProp } from "@react-navigation/native";
import { SharedStackParamList } from "./SharedNavParamList";
import CategoriesScreen from "@app/screens/Categories/CategoriesScreen";
import CategoryScreen from "@app/screens/Category/CategoryScreen";

interface CategoriesStackProps {}

type CategoriesStackParamList = {
  Categories: undefined;
  Category: {
    title: string;
  };
} & SharedStackParamList;

export type CategoriesStackNavProps<
  T extends keyof CategoriesStackParamList
> = {
  navigation: StackNavigationProp<CategoriesStackParamList, T>;
  route: RouteProp<CategoriesStackParamList, T>;
};

const Stack = createStackNavigator<CategoriesStackParamList>();


/**
 * Categories Stack contains CategoriesScreen and CategoryScreen
 *
 * @return {*} 
 */
export const CategoriesStack: React.FC<CategoriesStackProps> = () => {
  return (
    <Stack.Navigator >
      <Stack.Screen
        name="Categories"
        component={CategoriesScreen}
      ></Stack.Screen>
        <Stack.Screen
        name="Category"
        component={CategoryScreen}
      ></Stack.Screen>
    </Stack.Navigator>
  );
};
