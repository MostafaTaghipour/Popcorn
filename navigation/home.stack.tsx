import React from "react";
import {
  createStackNavigator,
  StackNavigationProp,
} from "@react-navigation/stack";
import { RouteProp } from "@react-navigation/native";
import { SharedStackParamList } from "./SharedNavParamList";
import HomeScreen from "@app/screens/Home/HomeScreen";
import SearchScreen from "@app/screens/Search/SearchScreen";

interface HomeStackProps {}

type HomeStackParamList = {
  Home: undefined;
  Search: undefined;
} & SharedStackParamList;

export type HomeStackNavProps<T extends keyof HomeStackParamList> = {
  navigation: StackNavigationProp<HomeStackParamList, T>;
  route: RouteProp<HomeStackParamList, T>;
};

const Stack = createStackNavigator<HomeStackParamList>();

/**
 * HomeStack contains HomeScreen and SearchScreen
 *
 * @return {*} 
 */
export const HomeStack: React.FC<HomeStackProps> = ({
  navigation,
  route,
}: any) => {
  navigation.setOptions({
    tabBarVisible: route.state ? (route.state.index > 0 ? false : true) : null,
  });
  return (
    <Stack.Navigator >
      <Stack.Screen name="Home" component={HomeScreen}></Stack.Screen>
      <Stack.Screen name="Search" component={SearchScreen}></Stack.Screen>
    </Stack.Navigator>
  );
};
