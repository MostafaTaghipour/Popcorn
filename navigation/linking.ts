import { LinkingOptions } from "@react-navigation/native/lib/typescript/src/types";
/**
 * If you are using universal links, you need to add your domain to the prefixes.
 * https://reactnavigation.org/docs/deep-linking/
 * 
 *  @type {*} */
const options: LinkingOptions = {
  prefixes: ["https://popcorn.com", "popcorn://"],
  // config: {
  // },
};
export default options;
