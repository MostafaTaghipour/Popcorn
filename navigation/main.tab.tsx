import React, { useContext } from "react";
import { createBottomTabNavigator } from "@react-navigation/bottom-tabs";
import { Feather } from "@expo/vector-icons";
import { HomeStack } from "./home.stack";
import { LocalizationContext } from "@app/contexts/LocalizationContext";
import { CategoriesStack } from "./categories.stack";
import { SettingsStack } from "./settings.stack";

interface MainTabProps {}

type MainTabParamList = {
  Home: undefined;
  Categories: undefined;
  Settings: undefined;
};

const Tab = createBottomTabNavigator<MainTabParamList>();



/**
 * MainTabs contains HomeStack , CategoriesStack and Settings
 *
 * @param {*} {}
 * @return {*} 
 */
export const MainTabs: React.FC<MainTabProps> = ({}) => {
  const { t } = useContext(LocalizationContext);

  return (
    <Tab.Navigator tabBarOptions={{ showLabel: false }}>
      <Tab.Screen
        name="Home"
        component={HomeStack}
        options={{
          title: t("home.title"),
          tabBarIcon: ({ color }) => {
            return <Feather name="home" size={24} color={color} />;
          },
        }}
      ></Tab.Screen>
      <Tab.Screen
        name="Categories"
        component={CategoriesStack}
        options={{
          title: t("categories.title"),
          tabBarIcon: ({ color }) => {
            return <Feather name="grid" size={24} color={color} />;
          },
        }}
      ></Tab.Screen>
      <Tab.Screen
        name="Settings"
        component={SettingsStack}
        options={{
          title: t("settings.title"),
          tabBarIcon: ({ color }) => {
            return <Feather name="settings" size={24} color={color} />;
          },
        }}
      ></Tab.Screen>
    </Tab.Navigator>
  );
};
