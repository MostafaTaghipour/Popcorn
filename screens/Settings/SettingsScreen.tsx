import { StyleSheet, SafeAreaView, View } from "react-native";
import React, { useContext } from "react";
import { Surface, Text, Switch, TouchableRipple } from "react-native-paper";
import { LocalizationContext } from "@app/contexts/LocalizationContext";
import { ThemeContext } from "@app/contexts/ThemeContext";
import defaultTheme from "@app/theme/default.theme";
import darkTheme from "@app/theme/dark.theme";
import { Feather } from "@expo/vector-icons";
import { showNativeConfirm } from "@app/helpers/MessageHelper";
import { AuthContext } from "@app/contexts/AuthContext";
import { useDispatch } from "react-redux";
import { signOutAction } from "@app/store/auth/actions";

/**
 * in Settings screen we can change theme and logout if we already logged in
 *
 * @export
 * @return {*}
 */
export default function SettingsScreen() {
  const { t } = useContext(LocalizationContext);
  const { isDark, setTheme } = useContext(ThemeContext);
  const { isAuthenticated } = useContext(AuthContext);
  const dispatch = useDispatch();

  /**
   * change current theme
   *
   */
  const changeTheme = () => {
    setTheme(isDark ? defaultTheme : darkTheme);
  };

  /**
   * logout user if confirm by user
   *
   */
  const logout = () => {
    showNativeConfirm({
      message: t("settings.are_u_sure"),
      positiveButtonTitle: t("yes"),
      negativeButtonTitle: t("no"),
      onPositiveButtonPress: () => {
        dispatch(signOutAction());
      },
    });
  };
  return (
    <SafeAreaView style={styles.container}>
      <Surface style={styles.surface}>
        <View style={styles.row}>
          <Switch value={isDark} onValueChange={changeTheme} />
          <Text style={styles.label}>{t("settings.dark_mode")}</Text>
        </View>
      </Surface>

      {isAuthenticated && (
        <Surface style={styles.surface}>
          <TouchableRipple onPress={logout} testID="LogoutButton">
            <View style={styles.row}>
              <Feather name="power" size={32} color="gray" />
              <Text style={styles.label}>{t("settings.log_out")}</Text>
            </View>
          </TouchableRipple>
        </Surface>
      )}
    </SafeAreaView>
  );
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
  },
  surface: {
    marginTop: 24,
    elevation: 0,
  },
  row: {
    padding: 16,
    flexDirection: "row",
    alignItems: "center",
  },
  label: {
    marginStart: 16,
  },
});
