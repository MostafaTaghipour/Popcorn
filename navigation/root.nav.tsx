import React, { useContext } from "react";
import { NavigationContainer, RouteProp } from "@react-navigation/native";
import { MainTabs } from "./main.tab";
import linking from "./linking";
import { ThemeContext } from "@app/contexts/ThemeContext";
import { Provider as PaperProvider } from "react-native-paper";
import {
  createStackNavigator,
  StackNavigationProp,
  TransitionPresets,
} from "@react-navigation/stack";
import { SharedStackParamList } from "./SharedNavParamList";
import IntroScreen from "@app/screens/Intro/IntroScreen";
import AuthScreen from "@app/screens/Auth/AuthScreen";
import { StatusBar, Platform } from "react-native";
import { Feather } from "@expo/vector-icons";
import { useReduxSelector } from "@app/store";

type AppNavigatorParamList = {
  MainTab: undefined;
  Intro: undefined;
} & SharedStackParamList;

export type AppNavigatorProps<T extends keyof AppNavigatorParamList> = {
  navigation: StackNavigationProp<AppNavigatorParamList, T>;
  route: RouteProp<AppNavigatorParamList, T>;
};

const Stack = createStackNavigator<AppNavigatorParamList>();

/**
 * RootNavigator handle screen according to seenIntro flag
 *
 * @param {*} {}
 * @return {*}
 */
export function RootNavigator() {
  const { theme, isDark } = useContext(ThemeContext);
  const seenIntro = useReduxSelector((state) => state.ui.seenIntro);

  return (
    <NavigationContainer theme={theme} linking={linking}>
      <PaperProvider theme={theme}>
        {/* Status bar according to current theme */}
        <StatusBar
          barStyle={isDark ? "light-content" : "dark-content"}
          backgroundColor={isDark ? "black" : "white"}
        />
        <Stack.Navigator screenOptions={{ cardOverlayEnabled: !isDark }}>
          {seenIntro ? (
            <>
              <Stack.Screen
                name="MainTab"
                component={MainTabs}
                options={{ headerShown: false }}
              />
              <Stack.Screen
                name="Auth"
                component={AuthScreen}
                options={{
                  headerBackTitleVisible: false,
                  headerBackImage: () => (
                    <Feather name="x" size={32} color={theme.colors.primary} />
                  ),
                  gestureEnabled: true,
                  cardOverlayEnabled: true,
                  ...Platform.select({
                    ios: { ...TransitionPresets.ModalPresentationIOS },
                    android: { ...TransitionPresets.ModalTransition },
                  }),
                  headerShown: false,
                }}
              />
            </>
          ) : (
            <Stack.Screen
              name="Intro"
              component={IntroScreen}
              options={{ headerShown: false }}
            />
          )}
        </Stack.Navigator>
      </PaperProvider>
    </NavigationContainer>
  );
}
