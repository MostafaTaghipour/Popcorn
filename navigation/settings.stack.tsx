import React from "react";
import {
  createStackNavigator,
  StackNavigationProp,
} from "@react-navigation/stack";
import { RouteProp } from "@react-navigation/native";
import { SharedStackParamList } from "./SharedNavParamList";
import SettingsScreen from "@app/screens/Settings/SettingsScreen";

interface SettingsStackProps {}

type SettingsStackParamList = {
  Settings: undefined;
} & SharedStackParamList;

export type SettingsStackNavProps<T extends keyof SettingsStackParamList> = {
  navigation: StackNavigationProp<SettingsStackParamList, T>;
  route: RouteProp<SettingsStackParamList, T>;
};

const Stack = createStackNavigator<SettingsStackParamList>();

/**
 * SettingsStack contains SettingsScreen 
 *
 * @return {*} 
 */
export const SettingsStack: React.FC<SettingsStackProps> = ({
  navigation,
  route,
}: any) => {
  navigation.setOptions({
    tabBarVisible: route.state ? (route.state.index > 0 ? false : true) : null,
  });

  return (
    <Stack.Navigator>
      <Stack.Screen name="Settings" component={SettingsScreen}></Stack.Screen>
    </Stack.Navigator>
  );
};
