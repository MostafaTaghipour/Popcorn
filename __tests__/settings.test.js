import React from "react";
import SettingsScreen from "../screens/Settings/SettingsScreen";
import { MockedNavigator } from "../navigation/MockedNavigator";
import configureStore from "redux-mock-store";
import { render } from "react-native-testing-library";

/**
 * test SettingsScreen rendering
 */
describe("Settings render", () => {
  let initialState;
  let store;

  /**
   * initialize mock redux store
   */
  beforeAll(() => {
    const middlewares = [];
    const mockStore = configureStore(middlewares);

    // Initialize mockstore
    initialState = {
      auth: {
        token: "TOKEN",
      },
    };
    store = mockStore(initialState);
  });

  /**
   * Test SettingsScreen snapshot
   */
  test("Settings snapshot", () => {
    const snap = render(
      <MockedNavigator component={SettingsScreen} store={store} />
    ).toJSON();
    expect(snap).toMatchSnapshot();
  });


  /**
   * If user is logged in LogoutButton must be shown
   */
  it("Testing LogoutButton visibility", () => {
    const tree = render(
      <MockedNavigator component={SettingsScreen} store={store} />
    );

    expect(tree.getAllByTestId("LogoutButton").length).toBe(1);
  });
});
