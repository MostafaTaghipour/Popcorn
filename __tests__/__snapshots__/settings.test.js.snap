// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Settings render Settings snapshot 1`] = `
<View
  style={
    Object {
      "flex": 1,
    }
  }
>
  <RNCSafeAreaView
    onInsetsChange={[Function]}
    style={
      Object {
        "flex": 1,
      }
    }
  >
    <View
      onLayout={[Function]}
      style={
        Object {
          "flex": 1,
        }
      }
    >
      <View
        pointerEvents="box-none"
        style={
          Object {
            "bottom": 0,
            "left": 0,
            "position": "absolute",
            "right": 0,
            "top": 0,
          }
        }
      >
        <View
          accessibilityElementsHidden={false}
          closing={false}
          gestureVelocityImpact={0.3}
          importantForAccessibility="auto"
          onClose={[Function]}
          onGestureBegin={[Function]}
          onGestureCanceled={[Function]}
          onOpen={[Function]}
          onTransitionStart={[Function]}
          pointerEvents="box-none"
          style={
            Object {
              "bottom": 0,
              "left": 0,
              "position": "absolute",
              "right": 0,
              "top": 0,
            }
          }
          transitionSpec={
            Object {
              "close": Object {
                "animation": "spring",
                "config": Object {
                  "damping": 500,
                  "mass": 3,
                  "overshootClamping": true,
                  "restDisplacementThreshold": 0.01,
                  "restSpeedThreshold": 0.01,
                  "stiffness": 1000,
                },
              },
              "open": Object {
                "animation": "spring",
                "config": Object {
                  "damping": 500,
                  "mass": 3,
                  "overshootClamping": true,
                  "restDisplacementThreshold": 0.01,
                  "restSpeedThreshold": 0.01,
                  "stiffness": 1000,
                },
              },
            }
          }
        >
          <View
            pointerEvents="box-none"
            style={
              Object {
                "flex": 1,
                "marginTop": 64,
              }
            }
          >
            <View
              collapsable={false}
              onGestureHandlerEvent={[Function]}
              onGestureHandlerStateChange={[Function]}
              style={
                Object {
                  "flex": 1,
                  "transform": Array [
                    Object {
                      "translateX": 0,
                    },
                    Object {
                      "translateX": 0,
                    },
                  ],
                }
              }
            >
              <View
                pointerEvents="none"
                style={
                  Object {
                    "backgroundColor": "rgb(242, 242, 242)",
                    "bottom": 0,
                    "left": 0,
                    "position": "absolute",
                    "shadowColor": "#000",
                    "shadowOffset": Object {
                      "height": 1,
                      "width": -1,
                    },
                    "shadowOpacity": 0.3,
                    "shadowRadius": 5,
                    "top": 0,
                    "width": 3,
                  }
                }
              />
              <View
                style={
                  Array [
                    Object {
                      "flex": 1,
                      "overflow": "hidden",
                    },
                    Array [
                      Object {
                        "backgroundColor": "rgb(242, 242, 242)",
                      },
                      undefined,
                    ],
                  ]
                }
              >
                <View
                  style={
                    Object {
                      "flex": 1,
                      "flexDirection": "column-reverse",
                    }
                  }
                >
                  <View
                    style={
                      Object {
                        "flex": 1,
                      }
                    }
                  >
                    <RCTSafeAreaView
                      emulateUnlessSupported={true}
                      style={
                        Object {
                          "flex": 1,
                        }
                      }
                    >
                      <View
                        style={
                          Object {
                            "backgroundColor": "#ffffff",
                            "elevation": 0,
                            "marginTop": 24,
                          }
                        }
                      >
                        <View
                          style={
                            Object {
                              "alignItems": "center",
                              "flexDirection": "row",
                              "padding": 16,
                            }
                          }
                        >
                          <RCTSwitch
                            accessibilityRole="button"
                            onChange={[Function]}
                            onResponderTerminationRequest={[Function]}
                            onStartShouldSetResponder={[Function]}
                            onTintColor="#03dac4"
                            style={
                              Object {
                                "height": 31,
                                "width": 51,
                              }
                            }
                            tintColor=""
                            value={false}
                          />
                          <Text
                            style={
                              Array [
                                Object {
                                  "color": "#000000",
                                  "fontFamily": "System",
                                  "fontWeight": "400",
                                  "textAlign": "left",
                                },
                                Object {
                                  "marginStart": 16,
                                },
                              ]
                            }
                          >
                            
                          </Text>
                        </View>
                      </View>
                      <View
                        style={
                          Object {
                            "backgroundColor": "#ffffff",
                            "elevation": 0,
                            "marginTop": 24,
                          }
                        }
                      >
                        <View
                          accessible={true}
                          focusable={true}
                          isTVSelectable={true}
                          onClick={[Function]}
                          onResponderGrant={[Function]}
                          onResponderMove={[Function]}
                          onResponderRelease={[Function]}
                          onResponderTerminate={[Function]}
                          onResponderTerminationRequest={[Function]}
                          onStartShouldSetResponder={[Function]}
                          style={
                            Array [
                              false,
                              undefined,
                            ]
                          }
                          testID="LogoutButton"
                        >
                          <View
                            style={
                              Object {
                                "alignItems": "center",
                                "flexDirection": "row",
                                "padding": 16,
                              }
                            }
                          >
                            <Text />
                            <Text
                              style={
                                Array [
                                  Object {
                                    "color": "#000000",
                                    "fontFamily": "System",
                                    "fontWeight": "400",
                                    "textAlign": "left",
                                  },
                                  Object {
                                    "marginStart": 16,
                                  },
                                ]
                              }
                            >
                              
                            </Text>
                          </View>
                        </View>
                      </View>
                    </RCTSafeAreaView>
                  </View>
                </View>
              </View>
            </View>
          </View>
        </View>
      </View>
    </View>
    <View
      pointerEvents="box-none"
      style={
        Object {
          "left": 0,
          "position": "absolute",
          "right": 0,
          "top": 0,
        }
      }
    >
      <View
        accessibilityElementsHidden={false}
        importantForAccessibility="auto"
        onLayout={[Function]}
        pointerEvents="box-none"
        style={null}
      >
        <View
          pointerEvents="none"
          style={
            Object {
              "bottom": 0,
              "left": 0,
              "opacity": 1,
              "position": "absolute",
              "right": 0,
              "top": 0,
            }
          }
        >
          <View
            style={
              Object {
                "backgroundColor": "rgb(255, 255, 255)",
                "borderBottomColor": "rgb(224, 224, 224)",
                "flex": 1,
                "shadowColor": "rgb(224, 224, 224)",
                "shadowOffset": Object {
                  "height": 0.5,
                  "width": 0,
                },
                "shadowOpacity": 0.85,
                "shadowRadius": 0,
              }
            }
          />
        </View>
        <View
          pointerEvents="box-none"
          style={
            Object {
              "height": 64,
              "maxHeight": undefined,
              "minHeight": undefined,
              "opacity": undefined,
              "transform": undefined,
            }
          }
        >
          <View
            pointerEvents="none"
            style={
              Object {
                "height": 20,
              }
            }
          />
          <View
            pointerEvents="box-none"
            style={
              Object {
                "alignItems": "center",
                "flex": 1,
                "flexDirection": "row",
                "justifyContent": "center",
              }
            }
          >
            <View
              pointerEvents="box-none"
              style={
                Object {
                  "marginHorizontal": 16,
                  "opacity": 1,
                }
              }
            >
              <Text
                accessibilityRole="header"
                numberOfLines={1}
                onLayout={[Function]}
                style={
                  Object {
                    "color": "rgb(28, 28, 30)",
                    "fontSize": 17,
                    "fontWeight": "600",
                  }
                }
              >
                MockedScreen
              </Text>
            </View>
          </View>
        </View>
      </View>
    </View>
  </RNCSafeAreaView>
</View>
`;
